### PR TITLE
[ENC-2013] Add tracing for tests

### PR DESCRIPTION
**Pull request by DomBlack**
**This PR must be merged at the same time as encoredev/go#27**

This commit adds tracing for tests. When tests are run, traces will be captured and stored like traces for API requests, which are exposed to the user via the Trace's view on the local dashboard.

## Screenshots of UI

image

image

image

### CODE REVIEW

File: cli/daemon/engine/trace2/sqlite/write.go [hunk 1]
```diff
@@ -203,6 +222,26 @@ func (s *Store) updateSpanEndIndex(ctx context.Context, meta *trace2.Meta, ev *t
 		return nil
 	}
 
+	if msg := end.GetTest(); msg != nil {
+		// TODO: record skipped test spans
```
Comment (eandre) [inline cli/daemon/engine/trace2/sqlite/write.go:226]: What does this mean?

File: pkg/traceparser/parser.go [hunk 2]
```diff
@@ -268,9 +272,16 @@ func (tp *traceParser) requestSpanStart() *tracepb2.SpanStart {
 				RequestPayload:   tp.ByteString(),
 				ExtCorrelationId: ptrOrNil(tp.String()),
 				Uid:              ptrOrNil(tp.String()),
+				Mocked:           false, // introduced in version 101
```
Comment (eandre) [inline pkg/traceparser/parser.go:275]: 101?

Comment (eandre) [inline pkg/traceparser/parser.go:275]: Nicer to do it inline: `(func() bool { ... })()`; that way we don't need the type assertion below

File: runtimes/go/appruntime/shared/reqtrack/trace_stream.go [hunk 3]
```diff
@@ -46,9 +47,17 @@ func (l *lazyTraceInit) Logger() trace2.Logger {
 	return l.log
 }
 
+func (l *lazyTraceInit) WaitForStreamSent() {
+	l.sent.Wait()
```
Comment (eandre) [inline runtimes/go/appruntime/shared/reqtrack/trace_stream.go:51]: This will return immediately if `l.initStream` hasn't been called. Is that intended?

Comment (DomBlack) [reply runtimes/go/appruntime/shared/reqtrack/trace_stream.go:51]: initStream is always called before WaitForStream; the only call site is:

```go
		op.trace.MarkDone()

		if blockOnTraceSend {
			op.trace.WaitForStreamSent()
		}
```

Where `MarkDone` is:

```go
func (l *lazyTraceInit) MarkDone() {
	l.log.MarkDone()
	l.initStream()
}
```

**Review by eandre — APPROVED** [review]

**Pull request merged**